Zonal face weights at constant lat

hongyuchen1030 opened the pull request:
<!--  The PR title should summarize the changes, for example "Add `Grid._build_face_dimension` function".
      Avoid non-descriptive titles such as "Addresses issue #229". -->

<!--  Replace XXX with the number of the issue that this PR will resolve. If this PR closed more than one,
      you may add a comma separated sequence-->
Closes #544 

## Overview
<!--  Please provide a few bullet points summarizing the changes in this PR. This should include
      points on any bug fixes, new functions, or other changes that have been made. -->
      
- The `_get_zonal_face_weights_at_constlat` method has been added to calculate the weight of each face at a constant latitude.
- The user can specify if the given edge is a great circle arc or a constant latitude one
- This method computes the perimeter of a circle at constant latitude, identifies overlapping intervals, and calculates the weight of each face based on the longitude bounds.
- Sweepline algorithm is used to calculate the overlapped intervals' weights
- Errors and edge cases are handled gracefully, with informative error messages and comments for future maintenance.

## Expected Usage
<!--  If you are adding a feature into the Internal API, please produce a short example of it in action.
      You may ignore this step if it is not applicable (comment out this section). -->
```Python
# Assuming uxarray is imported as ux
grid_path = "/path/to/grid.nc"
data_path = "/path/to/data.nc"

uxds = ux.open_dataset(grid_path, data_path)

# Replace `latitude_rad` with the desired latitude in radians
latitude_rad = 0.0  # Example value

        weight_df = _get_zonal_faces_weight_at_constLat(np.array([
            face_0_edge_nodes, face_1_edge_nodes, face_2_edge_nodes,
            face_3_edge_nodes
        ]),
                                                        0.01,
                                                        latlon_bounds,
                                                        is_directed=False)


```

## PR Checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. If an entire section doesn't
apply to this PR, comment it out or delete it. -->

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [x] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [x] User functions have been added to `docs/user_api/index.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

**PR Etiquette Reminders**
- This PR should be listed as a draft PR until you are ready to request reviewers

- After making changes in accordance with the reviews, re-request your reviewers

- Do *not* mark conversations as resolved if you didn't start them

- Do mark conversations as resolved *if you opened them* and are satisfied with the changes/discussion.
-->

hongyuchen1030 commented:
@philipc2 I have encountered the failure of the CI saying

```
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
Error: Process completed with exit code 1.
```

However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!

philipc2 commented:
> @philipc2 I have encountered the failure of the CI saying
> 
> ```
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> ================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
> Error: Process completed with exit code 1.
> ```
> 
> However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!

Interesting that it's only passing on MacOS. I don't believe it's a pandas version issue, but I would definitely update it to match with the version in the environment used for our CI tests.  

I gave the code a quick look through and couldn't figure out where that issue is occurring. I can give it a look later on.

hongyuchen1030 commented:
> > @philipc2 I have encountered the failure of the CI saying
> > ```
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_interval - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_GCA_constLat - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_face_weight_rad_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_equator - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - AttributeError: 'DataFrame' object has no attribute 'append'. Did you mean: '_append'?
> > ================ 5 failed, 149 passed, 1164 warnings in 47.34s =================
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, when I run on my local, all tests passed. My local pandas version is 1.5.2. I wonder if you have any clue about how to resolve this? Thank you very much!
> 
> Interesting that it's only passing on MacOS. I don't believe it's a pandas version issue, but I would definitely update it to match with the version in the environment used for our CI tests.
> 
> I gave the code a quick look through and couldn't figure out where that issue is occurring. I can give it a look later on.

Thank you very much for your help. As you can see, I have removed the `pandas.dataframe.append()` from my codes, and the current error messages on other system are that they can't find the intersection points, which is really abnormal since it's passing on MacOS. 

```
=========================== short test summary info ============================
FAILED test/test_integrate.py::TestFaceWeights::test_get_zonal_faces_weight_at_constLat_regular - ValueError: UXarray doesn't support concave face with intersections points as [3] currently, please modify your grids accordingly
```

hongyuchen1030 commented:
Note: Instead of using the `is_GCA_list`, use the flag as `is_unstructured`. If the `is_unstructured == true`, treat edge that has two endpoints that have the same latitude as the constant latitude

hongyuchen1030 commented:
Now the `is_latlonface` flag is added, which will be applied to all grids. But the `is_GCA` flag is still preserved and it can overwrite all other specification of the edge type.

Important note: The function will only accept convex face as input now. A more general version that take cares of the concave face will be implemented in the future. That means even if a polygon looks convex on a plane, it might look concave on the sphere, and our current function will still throw an exception for that

philipc2 commented on uxarray/grid/integrate.py:
```diff
@@ -0,0 +1,468 @@
+            overlap_flag = True
+
+        # Skip the dummy edge
+        if np.any(n1 == [INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]) or np.any(
+            n2 == [INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]
```
Can you turn `[INT_FILL_VALUE, INT_FILL_VALUE, INT_FILL_VALUE]` into a variable for readability?

philipc2 commented on uxarray/grid/integrate.py:
```diff
@@ -0,0 +1,468 @@
+    return weights_df
+
+
+def _get_zonal_face_interval(
```
This function is extremely long, with multiple if-else blocks that could be turned into helper functions. Can you turn some of those if-else blocks into helper functions?

philipc2 commented on uxarray/grid/integrate.py:
```diff
@@ -0,0 +1,468 @@
+"""uxarray grid module."""
```
```suggestion
```

philipc2 commented on uxarray/grid/integrate.py:
```diff
@@ -0,0 +1,468 @@
+"""uxarray grid module."""
+
```
I'm not sure if `integrate.py` is the best module name for this.

hongyuchen1030 replied:
Do you have any better suggestion? I call it "integrate" because it is described in our "Integration Algorithm" paper

erogluorhan reviewed (commented):
Re current CI failures:

With the commit entitled "[WIP](https://github.com/UXARRAY/uxarray/pull/555/commits/1598de2650bd9873482221ba774fe3e2b668be3a)" on the Feb 4th this year, several lines of codes seem to have been added. The following code block:

```
unique_intersection_lonlat = np.array(
                    [node_xyz_to_lonlat_rad(pt.tolist()) for pt in unique_intersection]
                )
>               unique_intersection_lonlat = unique_intersection_lonlat.append(
                    [2 * np.pi, latitude_rad]
                )
```

apparently creates a `numpy.ndarray` then attempts to `.append()` into that, which results in 

`AttributeError: 'numpy.ndarray' object has no attribute 'append'`

hongyuchen1030 commented:
> Re current CI failures:
> 
> With the commit entitled "[WIP](https://github.com/UXARRAY/uxarray/pull/555/commits/1598de2650bd9873482221ba774fe3e2b668be3a)" on the Feb 4th this year, several lines of codes seem to have been added. The following code block:
> 
> ```
> unique_intersection_lonlat = np.array(
>                     [node_xyz_to_lonlat_rad(pt.tolist()) for pt in unique_intersection]
>                 )
> >               unique_intersection_lonlat = unique_intersection_lonlat.append(
>                     [2 * np.pi, latitude_rad]
>                 )
> ```
> 
> apparently creates a `numpy.ndarray` then attempts to `.append()` into that, which results in
> 
> `AttributeError: 'numpy.ndarray' object has no attribute 'append'`
> 
> FYI: The last successful CI actions before the above commit seems to be [Merge branch 'main' into zonal_face_weights_at_constantlat](https://github.com/UXARRAY/uxarray/pull/555/commits/79bf078a90b47c2fa270ae20c75de70a7842ea31), which was on Jan 16.

Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.

erogluorhan commented:
> Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.

I think you intended to use `numpy.append(array-like1, array-like2,...)` rather than the one here (i.e. non-existent `numpy.ndarray.append()`?

hongyuchen1030 commented:
> > Thanks for your information. Yeah I also noticed that this was the cause. So this syntax is actually valid before. I assume it was the version upgrade that makes it corrupts. I am still working on fixing it since I have been working on other research project recently.
> 
> I think you intended to use `numpy.append(array-like1, array-like2,...)` rather than the one here (i.e. non-existent `numpy.ndarray.append()`?

I guessed the same thing before. But when I tried to fix it by using this `numpy.append(array-like1, array-like2,...)` , all my following codes and functions call corrupts due to incorrect array shape, which is not logical since my previous implementation passed all the test cases. I am still trying to figure out what I intended to do for that WIP commit. I had been actively working on this branch to test many current research ideas since it's been hanging there for a while.

hongyuchen1030 commented:
Hi @philipc2 @erogluorhan,

I have cleaned up and fixed the CI for this branch and it is mostly ready to go now. However, the read-doc failed is currently failing because the module `uxarray.Grid.grid_spec.`, which I didn't modify in this PR. I wonder if you can take a look into this? Thanks. Other than that, this PR is ready to be merged

> Running Sphinx v7.3.6
> making output directory... done
> [autosummary] generating autosummary for: api.rst, citation.rst, contributing.rst, examples.rst, examples/000-template.ipynb, examples/001-working-with-unstructured-grids.ipynb, examples/002-grid-topology.ipynb, examples/003-area-calc.ipynb, examples/004-working-with-mpas-grids.ipynb, examples/005-to-geodataframe-for-holoviz.ipynb, ..., examples/009-subsetting.ipynb, getting-started/freq-asked-questions.rst, getting-started/overview.rst, getting-started/quick-install.rst, index.rst, installation.rst, internal_api/index.rst, quickstart.rst, tutorials.rst, user_api/index.rst
> WARNING: [autosummary] failed to import uxarray.Grid.grid_spec.
> Possible hints:
> * PycodeError: no source found for module 'builtins'
> * AttributeError: type object 'Grid' has no attribute 'grid_spec'
> * ModuleNotFoundError: No module named 'uxarray.Grid'
> WARNING: [autosummary] failed to import uxarray.Grid.n_max_node_nodes.
> Possible hints:
> * AttributeError: type object 'Grid' has no attribute 'n_max_node_nodes'
> * PycodeError: no source found for module 'builtins'
> * ModuleNotFoundError: No module named 'uxarray.Grid'
> WARNING: [autosummary] failed to import uxarray.grid._slice_edge_indices.
> Possible hints:
> * ModuleNotFoundError: No module named 'uxarray.grid._slice_edge_indices'
> * ImportError: 
> * AttributeError: module 'uxarray.grid' has no attribute '_slice_edge_indices'
> WARNING: [autosummary] failed to import uxarray.grid._slice_face_indices.
> Possible hints:
> * ImportError: 
> * AttributeError: module 'uxarray.grid' has no attribute '_slice_face_indices'
> * ModuleNotFoundError: No module named 'uxarray.grid._slice_face_indices'
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._build_face_edges_connectivity.
> Possible hints:
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._build_face_edges_connectivity'; 'uxarray.grid.connectivity' is not a package
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_build_face_edges_connectivity'
> * ImportError: 
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._build_nNodes_per_face.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_build_nNodes_per_face'
> * ImportError: 
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._build_nNodes_per_face'; 'uxarray.grid.connectivity' is not a package
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._populate_face_edges_connectivity.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_populate_face_edges_connectivity'
> * ImportError: 
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._populate_face_edges_connectivity'; 'uxarray.grid.connectivity' is not a package
> WARNING: [autosummary] failed to import uxarray.grid.connectivity._populate_node_faces_connectivity.
> Possible hints:
> * AttributeError: module 'uxarray.grid.connectivity' has no attribute '_populate_node_faces_connectivity'
> * ModuleNotFoundError: No module named 'uxarray.grid.connectivity._populate_node_faces_connectivity'; 'uxarray.grid.connectivity' is not a package
> * ImportError:

philipc2 commented:
> Details

Hi @hongyuchen1030 

Thanks for the revisions! Giving this a look through now and I'll look into what's causing the RTD failures.

hongyuchen1030 commented:
> > Details
> 
> Hi @hongyuchen1030
> 
> Thanks for the revisions! Giving this a look through now and I'll look into what's causing the RTD failures.

Thank you very much! I just fixed a typo in the .rst file, I think the read-docs works now

philipc2 reviewed (approved):
This looks fantastic! Thanks for addressing all the comments! 

The documentation appears to be building fine. I believes the output you had were just warnings.

erogluorhan reviewed (approved):
This looks good to me. Thanks for addressing all the issues!
